[AMDGPU][AsmParser] Allow `v_writelane_b32` to use SGPR and M0 as source operands at the same time

shiltian opened the pull request:
Currently the asm parser takes `v_writelane_b32 v1, s13, m0` as illegal
instruction for pre-gfx11 because it uses two constant buses while the hardware
can only allow one. However, based on the comment of `AMDGPUInstructionSelector::selectWritelane`,
it is allowed to have M0 as lane selector and a SGPR used as SRC0 because the
lane selector doesn't count as a use of constant bus. In fact, codegen can already
generate this form, but this inconsistency is not exposed because the validation
of constant bus limitation only happens when paring an assembly but we don't have
a test case when both SGPR and M0 used as source operands for the instruction.

shiltian commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -3524,6 +3524,34 @@ bool AMDGPUAsmParser::validateConstantBusLimitations(
+  switch (Opcode) {
+  default:
+    break;
+  case V_WRITELANE_B32_e64_gfx11:
```
Is there any way that we don't have to list all variants here?

Sisyph replied:
This is basically the best way. SIInstrFlags bits can be used, but they are a limited resource. You could also create a tablegen mapping table such as is used for isVOPD, but it is probably overkill for this.

Sisyph commented on llvm/test/MC/AMDGPU/writelane_m0.s:
```diff
@@ -0,0 +1,10 @@
+// RUN: llvm-mc --triple=amdgcn --mcpu=gfx940 %s | FileCheck %s
+// RUN: llvm-mc --triple=amdgcn --mcpu=gfx1010 %s | FileCheck %s
+// RUN: llvm-mc --triple=amdgcn --mcpu=gfx1030 %s | FileCheck %s
+// RUN: llvm-mc --triple=amdgcn --mcpu=gfx1100 %s | FileCheck %s
```
Can you please add -show-encoding to all lines and add a gfx12 runline?

Sisyph reviewed (commented):
AMDGPUInstructionSelector::selectWritelane suggests that on targets where we can support 2 SGPRs per instruction (gfx10+) this is not a special case. That means' this is is only a behavior change for V_WRITELANE_B32_gfx6_gfx7 and V_WRITELANE_B32_vi. Can you please confirm that, and remove the newer versions of the instruction from your special case?

shiltian commented:
ping

shiltian commented:
ping +1

arsenm commented on llvm/test/MC/AMDGPU/writelane_m0.s:
```diff
@@ -0,0 +1,14 @@
+// RUN: llvm-mc --triple=amdgcn --mcpu=gfx1200 -show-encoding %s | FileCheck %s -check-prefix=GFX12
+
+.text
+  v_writelane_b32 v1, s13, m0
```
can you also add a machine verifier test with the MIR equivalents on different targets?

arsenm commented on llvm/test/CodeGen/AMDGPU/writelane_m0.mir:
```diff
@@ -0,0 +1,14 @@
+# RUN: llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx904 -verify-machineinstrs -run-pass=legalizer -o - %s | FileCheck %s
```
These tests should go in test/MachineVerifier, and only need -run-pass=none (don't even really need -verify-machineinstrs)

Sisyph commented on llvm/test/MC/AMDGPU/writelane_m0.s:
```diff
@@ -0,0 +1,14 @@
+// RUN: llvm-mc --triple=amdgcn --mcpu=gfx904 -show-encoding %s | FileCheck %s -check-prefix=GFX9
```
Looks like it needs a check line for gfx6/gfx7. Otherwise LGTM

shiltian committed:
[AMDGPU][AsmParser] Allow `v_writelane_b32` to use SGPR and M0 as source operands at the same time

Currently the asm parser takes `v_writelane_b32 v1, s13, m0` as illegal
instruction for pre-gfx11 because it uses two constant buses while the hardware
can only allow one. However, based on the comment of `AMDGPUInstructionSelector::selectWritelane`,
it is allowed to have M0 as lane selector and a SGPR used as SRC0 because the
lane selector doesn't count as a use of constant bus. In fact, codegen can already
generate this form, but this inconsistency is not exposed because the validation
of constant bus limitation only happens when paring an assembly but we don't have
a test case when both SGPR and M0 used as source operands for the instruction.

Sisyph reviewed (approved):
LGTM

arsenm commented on llvm/lib/Target/AMDGPU/AsmParser/AMDGPUAsmParser.cpp:
```diff
@@ -3514,6 +3514,24 @@ bool AMDGPUAsmParser::usesConstantBus(const MCInst &Inst, unsigned OpIdx) {
+  if (!LaneSelOp.isReg())
+    return false;
+  auto LaneSelReg = mc2PseudoReg(LaneSelOp.getReg());
+  if (LaneSelReg == M0 || LaneSelReg == M0_gfxpre11)
```
can fold this into return of bool expression
